components/base/sensorcontrolled/movestraight.go

```diff
@@ -0,0 +1,204 @@
+			if errors.Is(ctx.Err(), context.Canceled) {
+				sb.logger.Error(ctx.Err())
+				return nil
+			}
+			return ctx.Err()
+		case <-ticker.C:
+			var errDist float64
+
+			angVelDes, err := sb.calcHeadingControl(ctx, initialHeading)
+			if err != nil {
+				return err
+			}
+
+			if sb.position != nil {
+				errDist, err = sb.calcPositionError(ctx, distanceMm, initPos)
+				if err != nil {
+					return err
+				}
+			}
+
+			if errDist < moveStraightErrTarget {
+				return sb.Stop(ctx, nil)
+			}
+
+			linVelDes := calcLinVel(errDist, mmPerSec, slowDownDist)
+			if err != nil {
+				return err
+			}
+
+			// update velocity controller
+			if err := sb.updateControlConfig(ctx, linVelDes/1000.0, angVelDes); err != nil {
+				return err
+			}
+
+			// exit if the straight takes too long
+			if time.Since(startTime) > timeOut {
+				sb.logger.CWarn(ctx, "exceeded time for MoveStraightCall, stopping base")
+
+				return sb.Stop(ctx, nil)
+			}
+		}
+	}
+}
+
+// calculate the desired angular velocity to correct the heading of the base.
+func (sb *sensorBase) calcHeadingControl(ctx context.Context, initHeading float64) (float64, error) {
+	currHeading, err := sb.headingFunc(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	headingErr := initHeading - currHeading
+	headingErrWrapped := headingErr - (math.Floor((headingErr+180.)/(2*180.)))*2*180. // [-180;180)
+
+	return headingErrWrapped * headingGain, nil
+}
+
+// calcPositionError calculates the current error in position.
+// This results in the distance the base needs to travel to reach the goal.
+func (sb *sensorBase) calcPositionError(ctx context.Context, distanceMm int, initPos *geo.Point) (float64, error) {
+	pos, _, err := sb.position.Position(ctx, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	// the currDist will always return as positive, so we need the goal distanceMm to be positive
+	currDist := initPos.GreatCircleDistance(pos) * 1000000.
+	return math.Abs(float64(distanceMm)) - currDist, nil
+}
+
+// calcLinVel computes the desired linear velocity based on how far the base is from reaching the goal.
+func calcLinVel(errDist, mmPerSec, slowDownDist float64) float64 {
+	// have the velocity slow down when appoaching the goal. Otherwise use the desired velocity
+	linVel := errDist * mmPerSec / slowDownDist
+	absMmPerSec := math.Abs(mmPerSec)
+	if linVel > 0 && linVel > absMmPerSec {
+		return absMmPerSec
+	}
+	if linVel < 0 && linVel < -absMmPerSec {
+		return -absMmPerSec
+	}
+	return linVel
+}
+
+// calcSlowDownDist computes the distance at which the MoveStraight call should begin to slow down.
+// This helps to prevent overshoot when reaching the goal and reduces the jerk on the robot when the straight is complete.
+func calcSlowDownDist(distanceMm int) float64 {
+	slowDownDist := float64(distanceMm) * slowDownDistGain
+	if slowDownDist < -maxSlowDownDist {
+		return -maxSlowDownDist
+	}
+	if slowDownDist > maxSlowDownDist {
+		return maxSlowDownDist
+	}
+	return slowDownDist
+}
```

Comment:
this function is used to define if the robots velocity should be positive or negative from `distanceMm`. Not the biggest fan of using a 3rd variable to determine this, but `slowDownDist` also has functionality for having the robot slow down when we get close to the goal, so I am okay with this.